### PR TITLE
WIP: Make the container After Validation (Non Authed).

### PR DESCRIPTION
**Pull request by hibooboo2**

### CODE REVIEW

**Comment by hibooboo2** [comment]
* Ability for user to manage (CRUD) Docker registries via Rancher.
* Ability for users to select which Docker registry when launching containers.
* Ability to authenticate against Docker registries.
* Validate image if it is from index.docker.io. Don't validate otherwise in cattle
* Validate images and send up proper error from python agent to cattle when using nondocker repos.
* Auth comes from reg cred which has registry. 
* Be able to get all images that an account has used.

File: code/implementation/docker/storage/src/test/java/io/cattle/platform/docker/client/DockerImageTest.java [hunk 1]
```diff
@@ -25,9 +36,85 @@ public void testNamespace() {
...
+        DockerImage image = DockerImage.parse("quay.io/ubuntu-core:123");
+        assertEquals("quay.io",image.getServer());
+        assertEquals("ubuntu-core", image.getRepository());
+        assertNull(image.getNamespace());
+        assertEquals("123", image.getTag());
+    }
+
+    @Test
+    public void testLocalHost() {
+        DockerImage image = DockerImage.parse("localhost:5000/ibuildthecloud/ubuntu-core:123");
```
Comment (ibuildthecloud) [inline code/implementation/docker/storage/src/test/java/io/cattle/platform/docker/client/DockerImageTest.java]: Add a test for "localhost/ubuntu"

File: code/implementation/docker/storage/src/main/java/io/cattle/platform/docker/client/DockerImage.java [hunk 2]
```diff
@@ -1,44 +1,88 @@
...
+                    repo = id;
+                }else if (split2.length == 2){
+                    tag = split2[1];
+                    repo = split2[0];
+                }
+                break;
+            case 2:
+                String first = forwardSlash[0];
+                String[] second = forwardSlash[1].split(":");
+                if(first.contains(".") || first.contains(":") || first.contains("localhost")){
```
Comment (ibuildthecloud) [inline code/implementation/docker/storage/src/main/java/io/cattle/platform/docker/client/DockerImage.java]: first.equals("localhost")

File: code/implementation/docker/storage/src/test/java/io/cattle/platform/docker/client/DockerImageTest.java (hunk 1 above)
Comment (ibuildthecloud) [inline code/implementation/docker/storage/src/test/java/io/cattle/platform/docker/client/DockerImageTest.java]: add test for garbage/foo:1:2:3

File: code/implementation/docker/storage/src/main/java/io/cattle/platform/docker/storage/DockerStoragePoolDriver.java [hunk 3]
```diff
@@ -43,7 +43,9 @@ protected boolean populateExtenalImageInternal(StoragePool pool, String uuid, Im
             return false;
         }
 
-        dockerImage = dockerClient.lookup(dockerImage);
+//        if (dockerImage.getServer().equals("index.docker.io")) {
```
Comment (ibuildthecloud) [inline code/implementation/docker/storage/src/main/java/io/cattle/platform/docker/storage/DockerStoragePoolDriver.java]: remove

File: code/implementation/docker/storage/src/main/java/io/cattle/platform/docker/client/DockerImage.java [hunk 4]
```diff
@@ -1,44 +1,88 @@
...
+                }else {
+                    repo = forwardSlash[1];
+                    tag = "latest";
+                }
+                break;
+            case 3:
+                server = forwardSlash[0];
+                namespace = forwardSlash[1];
+                split2 = forwardSlash[2].split(":");
+                if (split2.length == 1){
```
Comment (ibuildthecloud) [inline code/implementation/docker/storage/src/main/java/io/cattle/platform/docker/client/DockerImage.java]: make consistent with L62

File: code/implementation/docker/storage/src/main/java/io/cattle/platform/docker/client/DockerImage.java [hunk 5]
```diff
@@ -74,7 +118,23 @@ public void setTag(String tag) {
     }
 
     public String getQualifiedName() {
-        if ( namespace == null ) {
+        if (server == "index.docker.io"){
```
Comment (ibuildthecloud) [inline code/implementation/docker/storage/src/main/java/io/cattle/platform/docker/client/DockerImage.java]: server.equals(....)

File: code/implementation/docker/storage/src/main/resources/META-INF/cattle/docker-storage/defaults.properties [hunk 6]
```diff
@@ -1,5 +1,6 @@
 docker.index.url=https://index.docker.io
+docker.index.server=index.docker.io
```
Comment (ibuildthecloud) [inline code/implementation/docker/storage/src/main/resources/META-INF/cattle/docker-storage/defaults.properties]: Remove if not used

Comment (ibuildthecloud) [inline code/implementation/docker/storage/src/main/resources/META-INF/cattle/docker-storage/defaults.properties]: Add to code/implementation/docker/storage/src/test/resources/config.properties for unit tests to work

File: resources/content/db/changelog.xml [hunk 7]
```diff
@@ -28,5 +28,6 @@
     <include file="db/core-021.xml"/>
     <include file="db/core-022.xml"/>
     <include file="db/core-023.xml"/>
+    <include file="db/core-024.xml"/>
```
Comment (ibuildthecloud) [inline resources/content/db/changelog.xml]: rebase

File: resources/content/schema/base/registry.json [hunk 8]
```diff
@@ -0,0 +1,9 @@
+{
+  "resourceFields": {
+    "serverAddress": {
+      "required" : true,
+      "nullable" : false,
```
Comment (ibuildthecloud) [inline resources/content/schema/base/registry.json]: minLength : 1

File: resources/content/schema/base/registry.json [hunk 9]
```diff
@@ -0,0 +1,9 @@
+{
+  "resourceFields": {
+    "serverAddress": {
+      "required" : true,
```
Comment (ibuildthecloud) [inline resources/content/schema/base/registry.json]: minLength : 1

File: resources/content/schema/base/registryCredential.json [hunk 10]
```diff
@@ -0,0 +1,15 @@
+{
+  "resourceFields": {
+    "email": {
```
Comment (ibuildthecloud) [inline resources/content/schema/base/registryCredential.json]: minLength : 1

File: resources/content/schema/base/registryCredential.json [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+      "required" : true,
+      "nullable" : false,
+      "type" : "string"
+    },
+    "storagePoolId": {
+      "required" : true,
+      "nullable" : false,
+      "type" : "reference[storagePool]",
+      "kind" : "registry"
```
Comment (ibuildthecloud) [inline resources/content/schema/base/registryCredential.json]: ?? i don't think this does anything

File: resources/content/schema/base/registryCredential.json [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+  "resourceFields": {
+    "email": {
+      "required" : true,
+      "nullable" : false,
+      "type" : "string"
+    },
+    "storagePoolId": {
+      "required" : true,
+      "nullable" : false,
+      "type" : "reference[storagePool]",
```
Comment (ibuildthecloud) [inline resources/content/schema/base/registryCredential.json]: reference[registry]

File: resources/content/schema/user/user-auth.json [hunk 13]
```diff
@@ -148,6 +149,14 @@
         "register.key":"cr",
         "register.secretKey" : "r",
 
+        "registry": "cr",
+        "registry.serverAddress": "cr",
+
+        "registryCredential": "cr",
+        "registryCredential.storagePoolId": "cr",
+        "registryCredential.email": "cr",
+        "storagePool": "r",
```
Comment (ibuildthecloud) [inline resources/content/schema/user/user-auth.json]: I like alphabetical :)

File: tests/integration/cattletest/core/test_docker.py [hunk 14]
```diff
@@ -71,7 +71,7 @@ def test_docker_create_only(admin_client, super_client, docker_context):
     assert image.data.dockerImage.repository == 'helloworld'
     assert image.data.dockerImage.namespace == 'ibuildthecloud'
     assert image.data.dockerImage.tag == 'latest'
-    assert image.data.dockerImage.id is not None
+    # assert image.data.dockerImage.id is not None
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_docker.py]: remove

File: code/implementation/docker/storage/src/test/java/io/cattle/platform/docker/client/DockerImageTest.java (hunk 1 above)
Comment (hibooboo2) [inline code/implementation/docker/storage/src/test/java/io/cattle/platform/docker/client/DockerImageTest.java]: Addressed with line 139

File: code/implementation/docker/storage/src/main/java/io/cattle/platform/docker/client/DockerImage.java (hunk 4 above)
Comment (hibooboo2) [inline code/implementation/docker/storage/src/main/java/io/cattle/platform/docker/client/DockerImage.java]: Done.

**Pull request closed without merging**